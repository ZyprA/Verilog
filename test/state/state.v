`define IDLE 2'b00
`define FETCH 2'b01
`define EXEC 2'b10

module state(clk, reset, run, halt, cs);
    input clk, reset, run, halt;
    output reg [1:0] cs;

    always @ (posedge clk or negedge reset)
        if (!reset) cs <= `IDLE;
        else
            case (cs)
                `IDLE: if(run) cs <= `FETCH;
                `FETCH: cs <= `EXEC;
                `EXEC: if (halt) cs <= `IDLE;
                else cs <= `FETCH;
                default: cs <= 2'bxx;
            endcase
endmodule

module state_sim;
    reg clk, reset, run, halt;
    wire [1:0] cs;

    state state_inst(clk, reset, run, halt, cs);

    initial begin
        $dumpfile("state.vcd");
        $dumpvars(0, state_sim);
    end

    initial begin
        clk = 0;
        forever #50 clk = ~clk;
    end

    initial begin
        reset = 0;
        run = 0;
        halt = 0;
        #100 reset = 1;
        #100 run = 1;
        #100 run = 0;
        #100 halt = 1;
        #100 halt = 0;
        #100 $finish;
    end
endmodule