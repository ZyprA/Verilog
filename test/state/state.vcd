$date
	Sun Feb  2 20:17:39 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module state_sim $end
$var wire 2 ! cs [1:0] $end
$var reg 1 " clk $end
$var reg 1 # halt $end
$var reg 1 $ reset $end
$var reg 1 % run $end
$scope module state_inst $end
$var wire 1 " clk $end
$var wire 1 # halt $end
$var wire 1 $ reset $end
$var wire 1 % run $end
$var reg 2 & cs [1:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b0 &
0%
0$
0#
0"
b0 !
$end
#50
1"
#100
0"
1$
#150
1"
#200
0"
1%
#250
b1 !
b1 &
1"
#300
0"
0%
#350
b10 !
b10 &
1"
#400
0"
1#
#450
b0 !
b0 &
1"
#500
0"
0#
#550
1"
#600
0"
