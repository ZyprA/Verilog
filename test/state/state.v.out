#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x11e609430 .scope module, "state_sim" "state_sim" 2 21;
 .timescale 0 0;
v0x11e61f360_0 .var "clk", 0 0;
v0x11e61f410_0 .net "cs", 1 0, v0x11e61f020_0;  1 drivers
v0x11e61f4a0_0 .var "halt", 0 0;
v0x11e61f570_0 .var "reset", 0 0;
v0x11e61f620_0 .var "run", 0 0;
S_0x11e607520 .scope module, "state_inst" "state" 2 25, 2 5 0, S_0x11e609430;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 1 "reset";
    .port_info 2 /INPUT 1 "run";
    .port_info 3 /INPUT 1 "halt";
    .port_info 4 /OUTPUT 2 "cs";
v0x11e60f4a0_0 .net "clk", 0 0, v0x11e61f360_0;  1 drivers
v0x11e61f020_0 .var "cs", 1 0;
v0x11e61f0c0_0 .net "halt", 0 0, v0x11e61f4a0_0;  1 drivers
v0x11e61f170_0 .net "reset", 0 0, v0x11e61f570_0;  1 drivers
v0x11e61f200_0 .net "run", 0 0, v0x11e61f620_0;  1 drivers
E_0x11e609b40/0 .event negedge, v0x11e61f170_0;
E_0x11e609b40/1 .event posedge, v0x11e60f4a0_0;
E_0x11e609b40 .event/or E_0x11e609b40/0, E_0x11e609b40/1;
    .scope S_0x11e607520;
T_0 ;
    %wait E_0x11e609b40;
    %load/vec4 v0x11e61f170_0;
    %nor/r;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.0, 8;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x11e61f020_0, 0;
    %jmp T_0.1;
T_0.0 ;
    %load/vec4 v0x11e61f020_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 2;
    %cmp/u;
    %jmp/1 T_0.2, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 2;
    %cmp/u;
    %jmp/1 T_0.3, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 2;
    %cmp/u;
    %jmp/1 T_0.4, 6;
    %pushi/vec4 3, 3, 2;
    %assign/vec4 v0x11e61f020_0, 0;
    %jmp T_0.6;
T_0.2 ;
    %load/vec4 v0x11e61f200_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.7, 8;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v0x11e61f020_0, 0;
T_0.7 ;
    %jmp T_0.6;
T_0.3 ;
    %pushi/vec4 2, 0, 2;
    %assign/vec4 v0x11e61f020_0, 0;
    %jmp T_0.6;
T_0.4 ;
    %load/vec4 v0x11e61f0c0_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.9, 8;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x11e61f020_0, 0;
    %jmp T_0.10;
T_0.9 ;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v0x11e61f020_0, 0;
T_0.10 ;
    %jmp T_0.6;
T_0.6 ;
    %pop/vec4 1;
T_0.1 ;
    %jmp T_0;
    .thread T_0;
    .scope S_0x11e609430;
T_1 ;
    %vpi_call 2 28 "$dumpfile", "state.vcd" {0 0 0};
    %vpi_call 2 29 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x11e609430 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x11e609430;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x11e61f360_0, 0, 1;
T_2.0 ;
    %delay 50, 0;
    %load/vec4 v0x11e61f360_0;
    %inv;
    %store/vec4 v0x11e61f360_0, 0, 1;
    %jmp T_2.0;
    %end;
    .thread T_2;
    .scope S_0x11e609430;
T_3 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x11e61f570_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x11e61f620_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x11e61f4a0_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x11e61f570_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x11e61f620_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x11e61f620_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x11e61f4a0_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x11e61f4a0_0, 0, 1;
    %delay 100, 0;
    %vpi_call 2 46 "$finish" {0 0 0};
    %end;
    .thread T_3;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "state.v";
